js/lib/AddressFmt.js

```diff
@@ -208,4 +209,209 @@ AddressFmt.prototype.format = function (address) {
 	return ret.replace(/\n+/g, '\n').trim();
 };
 
+
+/*
+ * Return true if this is an asian locale.
+ * @private
+ * @returns {boolean} true if this is an asian locale, or false otherwise
+ */
+function isAsianLocale(locale) {
+    return locale.language === "zh" || locale.language === "ja" || locale.language === "ko";
+}
+
+/*
+ * Invert the properties and values, filtering out all the values with numbers.
+ * @private
+ * @returns {Object} the inverted object
+ */
+function invertAndFilter(object) {
+    var ret = {};
+    for (var p in object) {
+        if (object.hasOwnProperty(p) && !object[p].match(/\d/)) {
+            ret[object[p]] = p;
+        }
+    }
+    
+    return ret;
+}
+
+/**
+ * Return information about the address format that can be used
+ * by UI builders to display a locale-sensitive set of input fields
+ * based on the current formatter's settings.<p>
+ *
+ * The object returned by this method is an array of address rows. Each
+ * row is itself an array which may have one to four address
+ * components in that row. Each address component is an object
+ * that contains a component property and a label to display
+ * with it. The label is written in the given locale, or the
+ * locale of this formatter if it was not given.<p>
+ *
+ * Optionally, if the address component is constrained to a
+ * particular pattern or to a fixed list of possible values, then
+ * the constraint rules are given in the "constraint" property.<p>
+ *
+ * If the constraint is that the address component must conform to a
+ * particular pattern, the regular expression that matches valid input
+ * is returned in "constraint". Often, it is only the postal code
+ * component that can be validated like this.<p>
+ *
+ * If the constraint is that the address component should be limited
+ * to a fixed list of values, then the constraint property will be
+ * set to an object that lists those values. The object maps codes for
+ * each valid value to labels to show in the UI for that value.
+ * The codes should not be shown to the user and are intended to
+ * represent the values in code. The labels are translated to the given
+ * locale or to the locale of this formatter if it was not given. For
+ * the most part, it is the region and country components that
+ * are constrained in this way. The list of values are sorted by the
+ * label where possible.<p>
+ *
+ * Here is what the result would look like for a US address:
+ * <pre>
+ * [
+ *   [{
+ *     "component": "streetAddress",
+ *     "label": "Street Address"
+ *   }],
+ *   [{
+ *     "component": "locality",
+ *     "label": "City"
+ *   },{
+ *     "component": "region",
+ *     "label": "State",
+ *     "constraint": {
+ *       "AL": "Alabama",
```

Comment:
JS objects don't preserve an order here, right? might be better to implement this as an array of label/value tuples

`[["AL", "Alabama"], ["AK", "Alaska"]] ...`

Reply:
Almost all engines in modern browsers preserve order (Firefox, Chrome, Edge) and also in recent versions of nodejs, Qt, and Rhino, but the spec doesn't mandate that you have to preserve in for..in loops, even in ES2015. So there is a de-facto standard but not an explicit standard. There are only a few older and lesser-used browsers that still do not preserve order. The rest of ilib already relies on order preservation, so this is a bigger issue than just this object. 

BUT, your comment here does bring up a good point. The order is not correct per locale! It is currently sorted according to the original English string. I think I will fix that in a subsequent change though to keep this one small.